Add 'state' parameter for alternatives

tprestegard opened the pull request:
Allow alternatives to be installed without being set as the current selection. Fixes #4543.

##### SUMMARY
Adds a `state` parameter to the alternatives module with the following behavior:
* When not set, or set to `selected`, the module behaves exactly as at present:
  * Every time this module runs, the alternative in question will be set as the current selection for the alternatives group
* When set to `present`, the alternative will be installed, but not set as the current selection for the group.
  * If the the alternative was previously set to `selected` and is now set to `present`, the group will be set to use auto mode.

This change preserves existing behavior by default and allows the following behavior if `state` is set to `present`:
* Auto mode can persist (presently manual mode is being forced)
* Manual user selection of alternative is preserved across successive Ansible runs.


##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
alternatives

##### ADDITIONAL INFORMATION
There is an example in the description of #4543.

A few questions:
* ~~Any better ideas for parameter names than `activate`?~~ Update: changed to `state`.
* Anyone else in favor of making `selected = False` the default behavior? At present, the PR preserves the existing default behavior, but to me it does not seem very useful.

felixfontein commented on changelogs/fragments/4557-alternatives-add-activate-parameter.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - alternatives - add `activate` parameter, which allows user to specify whether the alternative should be set as the active selection for its alternatives group.
```
```suggestion
  - alternatives - add ``activate`` parameter, which allows user to specify whether the alternative should be set as the active selection for its alternatives group (https://github.com/ansible-collections/community.general/issues/4543, https://github.com/ansible-collections/community.general/pull/4557).
```

felixfontein commented on changelogs/fragments/4557-alternatives-add-activate-parameter.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
```
This is a feature, not a bugfix:
```suggestion
minor_changes:
```

felixfontein commented on plugins/modules/system/alternatives.py:
```diff
@@ -41,6 +41,11 @@
+    description:
+      - Whether to set this alternative as the currently active selection.
+    type: bool
+    default: true
```
```suggestion
    default: true
    version_added: 4.8.0
```

felixfontein reviewed (commented):
Thanks for your contribution!

tprestegard commented:
> Looks sensible to me. How about adding a test for the new functionality (see tests/integration/targets/alternatives/tasks/)?
@felixfontein good idea, I'll take a look at doing that.

chriscroome commented:
> * Any better ideas for parameter names than activate?

The [manpage](https://manpages.debian.org/bullseye/dpkg/update-alternatives.1.en.html#Value:) has this for the query format:

> Value: currently-selected-alternative
>    The path of the currently selected alternative. It can also take the magic value none. It is used if the link doesn't exist.

Perhaps `selected` would be suitable?

tprestegard commented:
> Perhaps selected would be suitable?
Yes, that's great, I've changed it to `selected`.

felixfontein commented:
Hmm, maybe it shouldn't be a boolean parameter, but a parameter `state` with values `selected` (default) and `installed` instead?

After all, if you call this module with `selected: false`, I would expect that it selects *another* alternative so that the one I specified is no longer selected.

(`activate` was somewhat different, `activate: false` doesn't suggest to me that it is deactivated, only that it won't be activated in case it isn't already.)

tprestegard commented:
I'm OK with the boolean parameter; because of how `update-alternatives` works, I don't think `selected: false` necessarily implies that another alternative is selected - if no alternative is explicitly selected for a group, it will fall back to auto-selection based on priority.

Having said that, I don't feel too strongly, so I'm open to this change if you feel it's need.

chriscroome commented:
Good points, if `state` was used with one option being `selected`, then perhaps `present` would be better than `installed`?

This might be beyond the scope of this issue but in the future having `absent` as an option to `state`, to remove the alternative would be useful.

felixfontein commented:
> I'm OK with the boolean parameter; because of how `update-alternatives` works, I don't think `selected: false` necessarily implies that another alternative is selected

But Ansible is about ensuring state, not so much about doing actions. If `selected: false` isn't about ensuring that something is not selected, it's named wrong IMO. If this is a boolean option, something like `select_when_installing` would be a better name than `selected`.

> Good points, if `state` was used with one option being `selected`, then perhaps `present` would be better than `installed`?

Sure. `installed` was the first thing that came to my mind :-)

tprestegard commented:
OK, I like the idea of `state` with allowed values of `present` and `selected`, with future plans to add `absent` as another allowed value.

tprestegard commented:
To preserve the current behavior of this module, I would put `selected` as the default option `state`.  I am not a big fan of this behavior, as `present` seems to be a better default to me. What do others think? In general, do we try to stay away from changes like this to preserve backwards compatibility?

felixfontein commented:
In general, we do not break backwards compatibility if possible, and if we do we deprecate the old behavior and try to show deprecation warnings for a longer time before actually changing the behvaior. So if `state` is added, the default must be `selected`. It could be changed to `present` later (with the proper deprecation cycle), but that should only happen later (it's best to separate adding support for the new value, and deprecating the old one, by at least one major collection release to make migration easier for users).

felixfontein commented:
(See also #582 for our policy on deprecation and breaking changes.)

chriscroome commented:
I agree that having `selected` wouldn't be the best default if starting from scratch, however I don't think it is worth breaking backwards compatibility over this.

tprestegard commented:
Well, I updated the code to use the `state` parameter as discussed, added some handling for the case where an alternative is initially set as `selected` but then changed to `present`, and added some tests. Now attempting to debug the test issues.

tprestegard commented:
I can't seem to make the integration tests work locally unless I use the default test image. Is this the right invocation for testing with CentOS 7?
```
ansible-test integration --docker centos7 -v alternatives
```
With this one, I get an "unreachable" error, like it's having trouble SSH-ing to the container:
```
TASK [Gathering Facts] *************************************************************************************************************************************************************************************

fatal: [testhost]: UNREACHABLE! => {"changed": false, "msg": "Data could not be sent to remote host \"172.17.0.3\". Make sure this host can be reached over ssh: ssh: connect to host 172.17.0.3 port 22: Connection refused\r\n", "unreachable": true}
```

I've tried some other images and combinations of options with the `--target` flag and haven't gotten anything to work, other than the default test image with Python 3.9.  Most of the time, it seems like there is an issue with the target container failing to start. I'm probably doing something completely wrong... any suggestions?

felixfontein commented on plugins/modules/system/alternatives.py:
```diff
@@ -41,6 +41,16 @@
+  state:
+    description:
+      - C(present) - install the alternative (if not already installed), but do
+        not set it as the currently selected alternative for the group.
```
Maybe to clarify:
```suggestion
        not set it as the currently selected alternative for the group. If it already
        is the selected alternative for the group, it will not change that.
```

felixfontein commented:
(Looks like `centos7` also no longer works for me... `ubuntu2004` does, though. But that one does not fail.. I can reproduce it with a RHEL 8.4 remote though.)

felixfontein commented on tests/integration/targets/alternatives/tasks/tests_state.yml:
```diff
@@ -0,0 +1,71 @@
+    - { n: 2, priority: 70 }
+    - { n: 3, priority: 25 }
+
+- name: Ensure that the link group is in auto mode
```
How about printing some more info:
```suggestion
- name: Get contents of file
  command: cat {{ alternatives_dir }}/dummy

- name: Ensure that the link group is in auto mode
```
While running the tests on a RHEL 8.4 remote I was able to see:
```
/var/lib/alternatives$ cat dummy 
manual
/usr/bin/dummy

/usr/bin/dummy3
25
/usr/bin/dummy4
64
/usr/bin/dummy1
50
/usr/bin/dummy2
70
```
Having a `cat` directly before the command below would be more conclusive, though.

felixfontein commented:
I was able to fix it by adding `- "{{ alternatives_dir }}/dummy"` to the list of things to delete in `remove_links.yml`:
```.diff
 - name: remove links
   file:
     path: '{{ item }}'
     state: absent
   with_items:
+    - "{{ alternatives_dir }}/dummy"
     - /etc/alternatives/dummy
     - /usr/bin/dummy
```

russoz commented on plugins/modules/system/alternatives.py:
```diff
@@ -70,6 +87,15 @@
 from ansible.module_utils.basic import AnsibleModule
 
 
+class AlternativeState:
```
A bit of a nit-picking but I think both terms in the name should be plural:

- the name of the module is "alternatives" so we should be consistent
- within this class we have more than one state, so it should be "states". Line 109 `choices=AlternativeState.to_list()` sounds quite odd - to me at least.

Also I think we might want to consider prefixing that class name with a `_`, so that it doesn't become a public symbol. By being public, any change we make later on will require a long deprecation period.

felixfontein replied:
There are many arguments for and against using a plural vs. a singular name for an enumeration. Java and C# seem to prefer singular names (https://stackoverflow.com/a/15756009). I guess in the end it's a matter of taste.

> Also I think we might want to consider prefixing that class name with a `_`, so that it doesn't become a public symbol. By being public, any change we make later on will require a long deprecation period.

I wouldn't consider modules (as opposed to module_utils etc.) a public API. The [guidelines only talk about public/private module and plugin utils](https://github.com/ansible-collections/overview/blob/main/collection_requirements.rst#standards-for-developing-module-and-plugin-utilities). Maybe we should discuss this in community-topics...

russoz replied:
Fair enough. I reckon this is not a show stopper.

russoz commented on plugins/modules/system/alternatives.py:
```diff
@@ -141,18 +184,34 @@ def main():
+                )
+                changed = True
+
+        except subprocess.CalledProcessError as cpe:
```
Is this not covered by `check_rc=True`?

felixfontein replied:
Hmm, actually `subprocess.CalledProcessError` should never leave `module.run_command()`. The module has been catching it before, I guess both the old `catch` and the new one can be removed.

russoz replied:
That can also be done later.

russoz reviewed (approved):
LGTM

felixfontein commented:
@tprestegard thanks for your contribution!
@chriscroome @russoz thanks for reviewing!

If anyone wants to implement the improvements in a follow-up PR, they're welcome to do so :)

tprestegard commented:
@felixfontein thanks for guiding a new contributor through this process, I hope to be able to contribute more going forward!
